### PR TITLE
Issue #3830: Fix bug in DesignForExtension when order of annotations changes violation

### PR DESCRIPTION
**Pull request by MEZk**
#3830 

> @rnveach 
 As shown in the debug output above, do you know why DesignForExtensionCheck.hasAnnotation has to examine the contents of the method? Won't this lead to false positives for this method?
@romani 
it should not review content of method. modificators and annotations is enough for decision.

Fixed. 

```findFirstTokenByPredicate``` was reimplemented. The idea was taken from DetailAST#findFirstToken .

There is no need in looping through all ignored annotations and invoking of  ```hasAnnotation``` inside ```hasIgnoredAnnotation```  each time, that is why I decided to check whether ignored annotations set contains any of method annotations.

Diff reports:
http://mezk.github.io/i3830-DesignForExtension/diff

As you can see from the reports there are both false positives and false negatives in previous implementation of DesignForExtensionCheck. From my point of view, all of them are due to incorrect implementation of tree traversing algorithm in TokenUtils#findFirstTokenByPredicate.

### CODE REVIEW

**Comment by rnveach** [comment]
CI is passing, especially code coverage, but locally using EclEmma in eclipse I am seeing missing coverage for this check.
Running `clean cobertura:clean cobertura:cobertura` produces no results at all for this check. Its like it doesn't exist or is being excluded.

coverage1
coverage2

@romani @MEZk Can either confirm you are seeing the same issue? I haven't been able to identify what is causing the issue locally for me.
The method in the screenshot is what lead me to looking into because I was wondering why we needed some of the code for changes made to `findFirstTokenByPredicate`. We shouldn't need `null` check or above end of method line.

**Comment by MEZk** [comment]
@rnveach 
> Can either confirm you are seeing the same issue?

Confirm. Looks like cobertura does not instrument the class at all.

> The method in the screenshot is what lead me to looking into because I was wondering why we needed some of the code for changes made to findFirstTokenByPredicate. We shouldn't need null check or above end of method line.

Confirm. I removed both checks and regenerate reports locally. No NPEs. I will update diff reports a bit later.

**Comment by rnveach** [comment]
@MEZk Thanks for confirming.
This is not the only spot showing up missing coverage in this check.
`canBeOverridden` is missing coverage for `&& !ScopeUtils.isInInterfaceOrAnnotationBlock(methodDef)` and `&& !modifiers.branchContains(TokenTypes.ABSTRACT)`. It is not related to your changes.

I'll make an issue for this coverage problem.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/InputDesignForExtensionMultipleAnnotations.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        foo3();
+    }
+
+    @Deprecated
+    @Ignore
+    public void foo2() {
+        foo3();
+    }
+
+    private void foo3() {}
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/InputDesignForExtensionMultipleAnnotations.java]: please add test case for simple annotation usage for example "@Inject" , as you detected false-negative at https://mezk.github.io/i3830-DesignForExtension/diff/apache-struts/index.html#A2

Please add test cases for comments between annotations. 
Make test with annotations as single line with method signature.

Comment (MEZk) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/InputDesignForExtensionMultipleAnnotations.java]: Done

**Comment by MEZk** [comment]
I will regenerate diff report after https://github.com/checkstyle/checkstyle/pull/3845

**Comment by MEZk** [comment]
New reports.
https://mezk.github.io/i3830-DesignForExtension/diff/

The same amount of differences as in previous one.

**Comment by romani** [comment]
I am fine with changes, report looks good.

@rnveach , please review.

**Pull request merged**